truss/templates/server/model_wrapper.py

```diff
@@ -3,22 +3,83 @@
 import logging
 import sys
 import traceback
+from enum import Enum
 from pathlib import Path
-from typing import Dict, List
+from threading import Lock, Thread
+from typing import Dict, Union
 
-import kfserving
+import kserve
+from cloudevents.http import CloudEvent
+from kserve.grpc.grpc_predict_v2_pb2 import ModelInferRequest, ModelInferResponse
 from shared.secrets_resolver import SecretsResolver
 
+logger = logging.getLogger(__name__)
+
 MODEL_BASENAME = "model"
 
 
-class ModelWrapper(kfserving.KFModel):
+class ModelWrapper(kserve.Model):
+    class Status(Enum):
+        NOT_READY = 0
+        LOADING = 1
+        READY = 2
+        FAILED = 3
+
+    _config: dict
+    _model: object
+    _load_lock: Lock = Lock()
+    _predict_lock: Lock = Lock()
+    _status: Status = Status.NOT_READY
+
     def __init__(self, config: dict):
         super().__init__(MODEL_BASENAME)
         self._config = config
-        self._model = None
 
-    def load(self):
+    def load(self) -> bool:
+        if self.ready:
+            return self.ready
+
+        # if we are already loading, just pass; our container will return 503 while we're loading
+        if not self._load_lock.acquire(blocking=False):
+            return False
+
+        self._status = ModelWrapper.Status.LOADING
+
+        logger.info("Executing model.load()...")
```

Comment:
Should we put these under try block as well. These are unlikely to throw an exception but there doesn't seem any harm in moving them there.

Reply:
I tend to advocate for only things that have expected error cases to go into try blocks.  We don't expect logger to fail, so let's leave that outside it.